### PR TITLE
anago: Push images during staging and validate manifests during release

### PR DESCRIPTION
**Pull request by justaugustus**
#### What type of PR is this?

/kind bug regression
/priority critical-urgent

#### What this PR does / why we need it:

Establishes the following:

##### Mock stage

Push instead to a subdirectory of the core image staging repo: `gcr.io/k8s-staging-kubernetes/mock`

##### Mock release

Validate image manifests from the mock push location as one of the prerequisites.

##### Official stage

Build and push to `gcr.io/k8s-staging-kubernetes`.
(This means the images are now built against the real release commit instead of the mock one.)

##### Image promotion

After the official stage is complete:
- Generate a PR to handle the image promotion
- Ensure the images have been promoted to `k8s-artifacts-prod` endpoints **BEFORE** starting the official release

##### Official release

Validate image manifests have been promoted as one of the prerequisites.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/release/issues/1438

/assign @saschagrunert @tpepper
cc: @kubernetes/release-engineering 
/hold for testing

#### Special notes for your reviewer:

Note that the `release::docker::validate_remote_manifests` step comes before `push_git_objects`, so we avoid instances where a release is pushed to GitHub before the container images have been promoted.
ref: https://kubernetes.slack.com/archives/CJH2GBF7Y/p1595348328271900 / https://github.com/kubernetes/sig-release/issues/1156

#### Does this PR introduce a user-facing change?

```release-note
- anago: Push images during staging and validate manifests during release
-  lib/release: Cleanup container registry variables
  - Use GCRIO_PATH_PROD to construct KUBE_CROSS_REGISTRY
  - Rename GCRIO_PATH_TEST to GCRIO_PATH_STAGING
  - Add GCRIO_PATH_MOCK (gcr.io/k8s-staging-kubernetes/mock) for mocks
- lib/release: Use a mock location for images when running mock stages
```

### CODE REVIEW

File: lib/releaselib.sh [hunk 1]
```diff
@@ -33,9 +33,10 @@ readonly GCRIO_PATH_PROD="k8s.gcr.io"
 readonly GCRIO_PATH_PROD_GEO_ASIA="asia.gcr.io/k8s-artifacts-prod"
 readonly GCRIO_PATH_PROD_GEO_EU="eu.gcr.io/k8s-artifacts-prod"
 readonly GCRIO_PATH_PROD_GEO_US="us.gcr.io/k8s-artifacts-prod"
-readonly GCRIO_PATH_TEST="gcr.io/k8s-staging-kubernetes"
+readonly GCRIO_PATH_STAGING="gcr.io/k8s-staging-kubernetes"
+readonly GCRIO_PATH_MOCK="${GCRIO_PATH_STAGING}/mock"
 
-readonly KUBE_CROSS_REGISTRY="us.gcr.io/k8s-artifacts-prod/build-image"
+readonly KUBE_CROSS_REGISTRY="${GCRIO_PATH_PROD}/build-image"
```
Comment (xmudrii) [inline lib/releaselib.sh:39]: Is this missing a part of the URL?

Before it was `us.gcr.io/k8s-artifacts-prod/build-image`, but `GCRIO_PATH_PROD` is set to `k8s.gcr.io`, so this will equal to `k8s.gcr.io/build-image` at the end if I'm not mistaken.

Is this change intended?

Comment (xmudrii) [reply lib/releaselib.sh:39]: After thinking a bit more, this change makes sense. It's a bit early for me. :grimacing:

**Comment by justaugustus** [comment]
Running a test of this on the `release-1.18` branch:

```shell
TOOL_ORG=justaugustus TOOL_BRANCH=fix-staging-container-push krel gcbmgr --stage --branch release-1.18 --type official
```

```console
INFO Verifying repository state                   
INFO Repository is in clean state                 
INFO Found matching branch "fix-staging-container-push" 
INFO Found matching organization "justaugustus" and repository "release" in remote: a (git@github.com:justaugustus/release.git) 
INFO Verifying remote HEAD commit                 
INFO Got remote commit: 318b73683547f97ea604e3c88dbabbd1bce4b2a1 
INFO Verifying that remote commit is equal to the local one 
INFO Repository is up-to-date                     
INFO Running gcbmgr with the following options: &{Stage:true Release:false Stream:false Branch:release-1.18 ReleaseType:official BuildVersion: GcpUser: LastJobs:5 Repo:0xc000407cb0 Version:0xc000407cc0} 
INFO Build options: {  cloudbuild.yaml   kubernetes-release-test false false false   } 
INFO Retrieving Kubernetes release version for ci/latest on branch release-1.18 
INFO Retrieving Kubernetes build version from https://dl.k8s.io/ci/latest-1.18.txt... 
INFO Retrieved Kubernetes version: v1.18.7-rc.0.18+3ccee83c2885e0 
INFO Trying to get the kube-cross version for release-1.18... 
INFO Found the following kube-cross version: v1.13.14-1 
INFO Listing GCB substitutions prior to build submission... 
INFO PATCH_VERSION_TAG: 7                         
INFO RELEASE_BRANCH: release-1.18                 
INFO BUILD_AT_HEAD:                               
INFO BUILD_POINT: v1.18.7-rc.0.18-3ccee83c2885e0  
INFO BUILDVERSION: --buildversion=v1.18.7-rc.0.18+3ccee83c2885e0 
INFO MINOR_VERSION_TAG: 18                        
INFO NOMOCK_TAG:                                  
INFO TOOL_REPO: release                           
INFO GCP_USER_TAG: stephen-k8s-at-agst-us         
INFO TYPE: official                               
INFO KUBE_CROSS_VERSION: v1.13.14-1               
INFO MAJOR_VERSION_TAG: 1                         
INFO KUBERNETES_VERSION_TAG: 1.18.7               
INFO TOOL_ORG: justaugustus                       
INFO TOOL_BRANCH: fix-staging-container-push      
INFO TYPE_TAG: official                           
INFO NOMOCK:                                      
INFO Build directory: /home/augustus/go/src/k8s.io/release/gcb/stage 
INFO Config directory: /home/augustus/go/src/k8s.io/release/gcb/stage 
INFO Changing to build directory: /home/augustus/go/src/k8s.io/release/gcb/stage 
ID                                    CREATE_TIME                DURATION  SOURCE  IMAGES  STATUS
673748a0-cc12-45cd-a364-66d7a63099d8  2020-08-03T08:36:28+00:00  -         -       -       QUEUED
INFO: Refreshing access_token
Created [https://cloudbuild.googleapis.com/v1/projects/kubernetes-release-test/builds/673748a0-cc12-45cd-a364-66d7a63099d8].
Logs are available at [https://console.cloud.google.com/cloud-build/builds/673748a0-cc12-45cd-a364-66d7a63099d8?project=648026197307].
INFO: Display format: "
        table(
          id,
          createTime.date('%Y-%m-%dT%H:%M:%S%Oz', undefined='-'),
          duration(start=startTime,end=finishTime,precision=0,calendar=false,undefined="  -").slice(2:).join(""):label=DURATION,
          build_source(undefined="-"):label=SOURCE,
          build_images(undefined="-"):label=IMAGES,
          status
        )
      "
```

I chose an `official` release so we can verify the loop over `ORDERED_RELEASE_KEYS` works as intended for official and beta releases, and the `release-1.18`, so there's no possibility that we accidentally release 1.19 early :sweat_smile:

File: lib/releaselib.sh (hunk 1 above)
Comment (justaugustus) [reply lib/releaselib.sh:39]: No worries! :sweat_smile:
...and yes, `k8s.gcr.io/build-image` is the intended path.

**Comment by justaugustus** [comment]
Fails without https://github.com/kubernetes/release/pull/1455/commits/ac53f345c9668d1ef32f03737fb92a5ed7337126.
Rerunning at https://console.cloud.google.com/cloud-build/builds/51e10d1f-19a6-42ca-ae95-1a6c165dab38?project=648026197307.

**Comment by justaugustus** [comment]
/hold cancel
There are some other cleanups I see around the image logic, but I'll take that as a follow-up later today.
